### ISSUE TITLE
bocfel: Check for GLK_MODULE_SOUND instead of GARGLK in sound.c

### ISSUE THREAD

**Issue by townba** [garglk/garglk#372]
In bocfel's `sound.c`, it doesn't use any Gargoyle-specific functionality, but rather the standard Glk sound functions. It'd be great if the `#ifdef GARGLK` lines could be changed to `#ifdef GLK_MODULE_SOUND`. (I can make a pull request for this, but I wasn't sure there's not another upstream bocfel somewhere.)

```diff
diff --git a/sound.c b/sound.c
index 08591a9..4bedc5b 100644
--- a/sound.c
+++ b/sound.c
@@ -8,7 +8,7 @@
 
 #ifdef ZTERP_GLK
 #include <glk.h>
-#ifdef GARGLK
+#ifdef GLK_MODULE_SOUND
 #include <glkstart.h>
 #include <gi_blorb.h>
 static schanid_t sound_channel = NULL;
@@ -23,7 +23,7 @@ static schanid_t sound_channel = NULL;
  */
 void init_sound(bool story_is_blorb)
 {
-#ifdef GARGLK
+#ifdef GLK_MODULE_SOUND
   strid_t file = NULL;
 
   if(sound_loaded()) return;
@@ -65,7 +65,7 @@ void init_sound(bool story_is_blorb)
 
 bool sound_loaded(void)
 {
-#ifdef GARGLK
+#ifdef GLK_MODULE_SOUND
   return sound_channel != NULL;
 #else
   return false;
@@ -74,7 +74,7 @@ bool sound_loaded(void)
 
 void zsound_effect(void)
 {
-#ifdef GARGLK
+#ifdef GLK_MODULE_SOUND
   uint8_t repeats, volume;
   static uint32_t vols[8] = {
     0x02000, 0x04000, 0x06000, 0x08000,
```

**Comment by cspiegel** [comment]
Agreed that this ought to run under any Glk implementations that support it.  Two issues:

- Glk implementations aren't required to provide Blorb support, but the sound code assumes it exists (because it knows it does under Gargoyle).
  - Most _do_ provide Blorb support, though, so it'd be easy to just assume it exists but allow it to be disabled at compile time.
- The code also makes use of `glkunix_stream_open_pathname` to open the Blorb file. This file is only expected to exist on Unix implementations, as well as in Gargoyle. It won't work, for example, on WinGlk.
  - A first cut would be to only support on Unix/Gargoyle; and I am able to cross compile for Windows, so I'd just have to figure out how to open files with WinGlk.

In short, this isn't a horribly onerous task, but it's a bit more involved than just replacing `GARGLK` with `GLK_MODULE_SOUND`.

**Comment by townba** [comment]
Good point. My testing was with Blorb examples, then I worked with Glulxe to get it working with SND files alongside Glulx game file, but I didn't try SND files alongside Z-machine files.

I see the following:

* `glk_schannel_create` is only called when Bocfel's `init_sound` has a Blorb file. I think it might be okay to call `glk_schannel_create` even if we don't have a Blorb file as long as the Glk sound code is expected to do the right thing to find the source resources.
* The Glk sound code needs to find the sound resources, but it doesn't have any knowledge of where the game file is. I handled that in a build of Glk for Glulx by remembering the filenames in `glk_stream_open_pathname` and `glk_stream_open_file` (from the `fileref_t`), but Bocfel isn't opening the game file through that path, so that doesn't work. My Glk sound code looks at the locations of open streams to see if it can find sound resources, but it doesn't find anything.

It might be harder because from what I've seen, Glk doesn't have a way to give it a hint about where it might find resources if they're not in a Blorb. Any ideas about the best way to handle something like this?

**Comment by townba** [comment]
Looking at this a bit more, perhaps my code to open the sidecar SND%d files could use a similar approach to that used by `glk_fileref_create_by_name`. In glkterm and cheapglk, they use the current working directory ('.') to look for files. Rather than trying to be fancy, maybe that's what the Glk sound code should do when it doesn't have a Blorb.

**Comment by cspiegel** [comment]
It looks like Gargoyle supports SND resources if there's no Blorb map set.  So it may be worthwhile to try to play sounds even if Bocfel doesn't think a Blorb file exists, as you say, in the hopes that the Glk library will handle it.  Looking at Gargoyle's SDL sound code, though, it will allow a channel to be created unconditionally, failing only when sounds are played.  That's fine, except it looks like it'll display an error message (at least to stderr, so not in the main window) if it can't find the resource, including if there is no Blorb file and no SND files.

Bocfel supports being built without Glk support, which complicates matters somewhat.  It uses standard I/O routines to open files, and does its own Blorb handling. As such, it doesn't have any handle to the game file that Glk would understand.  That's fine on Gargoyle, because Gargoyle allows you to get a filename from a Glk file reference (e.g. when the user chooses a file from the file picker).  But by the time the game file is loaded into memory, it's just a pointer to the game file, not the Blorb file, so that a function like `glk_stream_open_memory` won't work to find the Blorb contents, even if the game file was a Blorb.

I need to keep Bocfel's I/O separate from Glk so that it can continue to be built without Glk, so I can't just start "properly" using Glk functions.  Even if you could get a handle to the Blorb file after initially opening the file via the file picker (and you can), it's also possible to specify the filename on the command line, and Glk is not able to open files by name like that.

I think the only real cross-Glk approach is to read the entire Blorb file into memory and process it with `glk_stream_open_memory`.  I'd probably want to add some `mmap`-style functionality to `osdep.c` though, because I don't particularly like the idea of loading files of unknown size into memory, nor do I want to cap the file size at some arbitrary value.

**Comment by townba** [comment]
I really respect its ability to use or not use Glk. No small feat.

One short-term approach for Glk might be to call `glkunix_set_base_file`. (Yeah, I know, it's another `glkunix_` function.) Gargoyle, glkterm, and cheapglk all have it, and I don't think it makes the library-support situation any worse. Gargoyle's sound code uses it to get a hint about where to find SND%d files (among other things), and my sound code could do the same. (Honestly, it should have already, but I only just realized I could use it.) (Unfortunately, it looks like only Gargoyle's libgarglkmain calls it when the file is passed on the command line, but oh well.)

If Bocfel calls Glk sound code even without a Blorb file, it gives the Glk sound code a chance to find the resources. If it also calls `glkunix_set_base_file`, it might increase the chances. I'll change my code to also check the working directory set by `glkunix_set_base_file`.

(FYI, with just a handful of `#ifdef` changes, Bocfel works beautifully with a build of glkterm with sound and style support. I'm sure you're not surprised.)

**Comment by cspiegel** [comment]
I just did a first cut by restricting sound loading code to Unix, which, as you mention, ought to be a superset of what exists now (that is, Gargoyle will continue to work, and now Unix-style Glk implementations will too).  The problem: nanoglk (possibly others) doesn't have `glkunix_set_base_file`.  So it can't just unconditionally be called on Unix, since there's no guarantee it exists.  Although to be honest, I don't suppose there's a guarantee that `glkunix_stream_open_pathname` will exist on all Unix-based Glk implementations either, although it does appear to...

However, that's not crucial to getting the sound code working generally, and I'll probably wind up adding a call to `glkunix_set_base_file` with a build option to disable it, since the "main" Unix Glk implementations do support it.  I want to be compatible with all Glk implementations, but I can also be pragmatic.

I actually spent time getting code working in a cross-platform fashion by `mmap`ing the Blorb file and using `glk_stream_open_memory`, only to discover Gargoyle doesn't support Blorb files loaded from memory, so there would have to be two separate Blorb-loading code paths.  Maybe someday.

For the moment, then, I'm sticking with Unix only.  I've got things working here using `GLK_MODULE_SOUND` instead of `GARGLK`, although the only sound-enabled Glk library I have currently is Gargoyle, so I can't fully test it.

@townba , it sounds like you've modified glktermw to support sound?  If so, once I'm comfortable with the changes here I can send you a copy of the updated Bocfel source to test against, or feel free to send me your code and I can test against it.

And I'm glad to hear that it's working well with your modified glkterm; I've tried hard to get Bocfel to work as far and wide as possible, but there's only so much I can test.  Thanks for the report!

**Comment by townba** [comment]
Yes, I've got a modified glkterm that supports sound (and styles, which could use similar `GARGLK`-to-`GLK_MODULE_GARGLKTEXT` #ifdef tweaks in Bocfel, but that's another story). I'd be happy to try whatever you've got. If *you'd* like to try it, it's at https://github.com/townba/glkterm in the `townba-sound` branch.

**Comment by cspiegel** [comment]
I've updated Bocfel to make use of both `GLK_MODULE_SOUND` and `GLK_MODULE_GARGLKTEXT`.  I tested it against the relevant branches you've got, and it seems to be working well.

I'm attaching a diff from Bocfel 1.0.1 to the current Git master, if you'd like to give it a run through.  It's got a fair number of other changes alongside the Glk ones mentioned, but nothing that should hinder testing.

I'll probably do a new Bocfel release soon that includes these changes, assuming all looks good.

[bocfel-update.diff.txt](https://github.com/garglk/garglk/files/4498293/bocfel-update.diff.txt)

**Comment by townba** [comment]
I just gave it a quick try. Looks great on both fronts!

To enable sound support, I did have to define `ZTERP_GLK_BLORB`, which is fine. FWIW, I think abandoning non-Blorb sound support is probably the right call. I just checked, and we're farther from the creation of Blorb than the creation of Blorb was from the first release of *Zork*.

**Comment by cspiegel** [comment]
> It might be harder because from what I've seen, Glk doesn't have a way to give it a hint about where it might find resources if they're not in a Blorb. Any ideas about the best way to handle something like this?

Sorry, I completely overlooked this.  Unix implementations should have `glkunix_set_base_file` which tells the Glk implementation where the game file is located.  I don't actually call this in Bocfel... but I will add it.  Gargoyle currently handles it by looking at `argv` before calling `glk_main`, assuming that the last argument is a filename.  Not perfect by any means, but pretty decent, for interpreters which don't give the information some other way.

And Windows Glk looks to have `winglk_set_resource_directory` which serves the same purpose.

I'll add these calls before release.

> To enable sound support, I did have to define `ZTERP_GLK_BLORB`, which is fine.

If you're using Bocfel's Makefile, it should be setting that automatically when Glk on Unix is being used; is that not happening for you?

**Comment by townba** [comment]
> I'll add these calls before release.

I added `glkunix_set_base_file(game_file);` at the end of `process_arguments`, and boom, it worked with separate SND%d files. So with something like that, it supports all three approaches:

* game.zblorb (game and sounds in one file)
* game.z5, game.blorb (game in one file, sounds in another)
* game.z5, SND3, SND4, … (game in one file, sounds in separate files)

> If you're using Bocfel's Makefile, it should be setting that automatically when Glk on Unix is being used; is that not happening for you?

Oops. I'm sorry. I should have tried that first! Yes, it works. (I have an Xcode project set up that I'm using for all this stuff.)

**Comment by townba** [comment]
Just found one tiny change needed in the patch. In `write_header`, it should be checking for `GLK_MODULE_GARGLKTEXT` instead of `GARGLK` to set the color bit in Flags-1 , like this:

```c
#ifdef ZTERP_GLK
    if(glk_gestalt(gestalt_Timer, 0)) flags1 |= FLAGS1_TIMED;
#ifdef GLK_MODULE_GARGLKTEXT
    if(zversion >= 5) flags1 |= FLAGS1_COLORS;
#endif
#else
    if(!zterp_os_have_style(STYLE_BOLD)) flags1 &= ~FLAGS1_BOLD;
    if(!zterp_os_have_style(STYLE_ITALIC)) flags1 &= ~FLAGS1_ITALIC;
    if(zversion >= 5 && zterp_os_have_colors()) flags1 |= FLAGS1_COLORS;
#endif

    if(zversion >= 5 && options.disable_color) flags1 &= ~FLAGS1_COLORS;
    if(options.disable_timed) flags1 &= ~FLAGS1_TIMED;
    if(options.disable_fixed) flags1 &= ~FLAGS1_FIXED;
  }
```

**Comment by cspiegel** [comment]
Ah, nice catch, thanks!

**Comment by cspiegel** [comment]
Bocfel 1.1 has been released with support for `GLK_MODULE_SOUND` among other updates.  Thank you again for pushing this issue!
